### PR TITLE
Chore: (Docs) Auto docs fixes

### PR DESCRIPTION
**Pull request by jonniebigodes**
With this pull request, the documentation was updated to clarify and fix some breaking changes regarding the restructuring of various sections.

What was done:
- Updated the table of contents to reflect the naming of the autodocs feature.
- Updated the frameworks & feature support framework to feature the changes for the upcoming Doc Blocks changes and other items left out (e.g., interaction testing).
- Polished the Autodocs page based on the feedback
- Polished MDX based on feedback
- Updated the snippets to address: #20913, #20091, #21007
 - Added the required snippets per the changes.

### CODE REVIEW

File: docs/writing-docs/autodocs.md [hunk 1]
```diff
@@ -190,8 +234,50 @@ Out of the box, Storybook has a set of components that you can use to customize
...
+
+</div>
+
+## Troubleshooting
+
+### The auto-generated documentation is not showing up in a monorepo setup
+
+Out of the box, Storybook's AutoDocs feature is built to generate documentation for your stories automatically. Nevertheless, if you're working with a monorepo setup (e.g., [`Yarn Workspaces`](https://yarnpkg.com/features/workspaces), [`pnpm Workspaces`](https://pnpm.io/workspaces)), you may run into issues where part of the documentation may not be generated for you. To help you troubleshoot those issues, we've prepared some recommendations that might help you.
+
+Update your imports statement to reference the component directly instead of the package's root. For example:
```
Comment (JReinhold) [inline docs/writing-docs/autodocs.md:249]:
```suggestion
Update your import statements to reference the component directly instead of the package's root. For example:
```

File: docs/writing-docs/mdx.md [hunk 2]
```diff
@@ -254,36 +285,77 @@ import Changelog from '../CHANGELOG.md';
...
+  "editor.minimap.enabled": false,
+  "editor.formatOnSave": true,
+  "editor.formatOnType": true,
+  "editor.formatOnPaste": true,
+  "editor.inlineSuggest.enabled": true,
+  "editor.bracketPairColorization.enabled": true,
+  "editor.guides.bracketPairs": "active",
+  "editor.stickyScroll.enabled": true,
+  "mdx.experimentalLanguageServer": true
+}
```
Comment (JReinhold) [inline docs/writing-docs/mdx.md:321]: It's only really the last setting that's relevant here, right?
```suggestion
{
  "mdx.experimentalLanguageServer": true
}
```

**Review by JReinhold — APPROVED** [review]
Looked through the changes and it looks good! Is there an easy way to preview the changes on the site?

File: docs/writing-docs/mdx.md (hunk 2 above)
Comment (jonniebigodes) [reply docs/writing-docs/mdx.md]: I'm good with the suggestion and have already applied it. I left it so that I was able to give some context on what the actual structure be.

File: docs/frameworks.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 module.exports = {
-  coreFrameworks: ['react', 'vue', 'angular', 'web-components'],
+  coreFrameworks: ['react', 'vue', 'angular', 'web-components', 'nextjs'],
```
Comment (kylegach) [inline docs/frameworks.js:2]: Why was this added? Doing so will result in `frontpage` adding it to the [list of core frameworks everywhere](https://github.com/storybookjs/frontpage/blob/main/site-metadata.js#L1-L5), including the FrameworkSelector menu.

File: docs/snippets/common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+
+```mdx
+{/* Button.mdx */}
+
+import { Meta } from '@storybook/blocks';
+
+<Meta title="Button" />
+
+# Definition
```
Comment (kylegach) [inline docs/snippets/common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx:9]: Would this be more realistic?

```suggestion
# Button

## Definition
```

Comment (kylegach) [reply docs/snippets/common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx:9]: Applies to many of these snippets.

File: docs/snippets/common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## Usage
+
+The component comes in different variants such as `primary`, `secondary`, `large` and `small` which you can use to alter the look and feel of the button.
+
+## Inputs
+
+Button has the following properties:
+
+- `primary` - If `true`, the button will have a primary style.
```
Comment (kylegach) [inline docs/snippets/common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx:25]: This feels like an anti-pattern, because the ArgTypes or Controls blocks are a _much_ better means of documenting these.

File: docs/writing-docs/autodocs.md [hunk 6]
```diff
@@ -55,16 +55,41 @@ By default, Storybook offers zero-config support for documentation and automatic
...
+<CodeSnippets
+  paths={[
+    'common/storybook-preview-auto-docs-custom-template-function.js.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+<div class="aside">
+💡 Internally, Storybook uses a similar implementation to generate the default template. See the Doc Blocks API reference if you want to learn more about how Doc Blocks work.
```
Comment (kylegach) [inline docs/writing-docs/autodocs.md:79]: Let's add a link for the API reference and coordinate this PR merging with https://github.com/storybookjs/storybook/pull/21113

File: docs/writing-docs/autodocs.md [hunk 7]
```diff
@@ -55,16 +55,41 @@ By default, Storybook offers zero-config support for documentation and automatic
...
+Going over the code snippet in more detail. When Storybook starts up, it will override the default template with the custom one composed of the following:
+
+1. A header with the component's metadata retrieved by the `Title`, `Subtitle`, and `Description` Doc Blocks.
+2. The first story defined in the file via the `Primary` Doc Block with a handy set of UI controls to zoom in and out of the component.
+3. An interactive table with all the relevant [`args`](../writing-stories/args.md) and [`argTypes`](../api/argtypes.md) defined in the story via the `Controls` Doc Block.
+4. A overview of the remaining stories via the `Stories` Doc Block.
+
+#### With MDX
+
+However, you're not restricted to defining a function to generate the documentation template. The `docs` [parameter](../writing-stories/parameters.md#global-parameters) also accepts a reference to a template written in MDX, but this comes with a caveat, depending on where you place the file in your project, it may inadvertently get picked up by Storybook's auto-generated documentation. We recommend placing the file in a custom directory outside the `src` directory to avoid this issue. For example:
```
Comment (kylegach) [inline docs/writing-docs/autodocs.md:92]: A few notes here:

1. Once available, "`docs` parameter" should link to the API reference, not generic parameters documentation.
2. I'm pretty sure I've seen @tmeasday mention that you must provide a React component now and can no longer use MDX for this.
3. Not all projects use a `src` directory. Better to reference the `stories` configuration and recommend placing it somewhere that won't match?
    - This is moot if I'm correct about (2)

File: docs/writing-docs/autodocs.md [hunk 8]
```diff
@@ -98,25 +123,44 @@ Storybook provides support for MDX 2, an open standard markup language, combinin
 
 <!-- prettier-ignore-end -->
 
-Create an MDX file for your component in the same directory as your stories and components to add your custom documentation. We recommend naming the file similar to your component (e.g., `Button.mdx`) for easy reference.
+Create an MDX file to add your custom documentation. Depending on how you want your documentation to render in the UI, you'll need to consider the following use cases.
+
+### Using the `Meta` Doc Block
```
Comment (kylegach) [inline docs/writing-docs/autodocs.md:128]: If this page is now titled _Automatic documentation and Storybook_, I don't understand why this entire section (_Setup custom documentation_) is located here and not as part of the _MDX_ page.

All of these examples demonstrate how to write custom MDX docs entries and have nothing to do with autoDocs.

File: docs/writing-docs/autodocs.md [hunk 9]
```diff
@@ -98,25 +123,44 @@ Storybook provides support for MDX 2, an open standard markup language, combinin
...
+    'common/storybook-auto-docs-baseline-example.custom-title.mdx.mdx',
+    'common/storybook-auto-docs-baseline-example.of-prop.mdx.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+### Using the File System
+
+However, providing the `Meta` Doc Block may not be required for certain use cases, such as standalone pages or even as guidelines for testing your components. In that case, you can safely omit it. Storybook will instead rely on the file's physical location to place the documentation in the sidebar, overriding any pre-existent auto-generated documentation with your own. For example:
```
Comment (kylegach) [inline docs/writing-docs/autodocs.md:145]: Could we reference and leverage the [existing autotitle docs](https://storybook.js.org/docs/7.0/react/configure/sidebar-and-urls#csf-30-auto-titles) when explaining this?

File: docs/snippets/common/storybook-main-fix-imports-autodocs-monorepo.js.mdx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  ],
+  framework: {
+    // Replace your-framework with the framework you are using (e.g., react-webpack5, vue3-webpack5)
+    name: '@storybook/your-framework',
+    options: {},
+  },
+  typescript: {
+    reactDocgen: 'react-docgen',
+    skipBabel: true,
+    check: false,
```
Comment (kylegach) [inline docs/snippets/common/storybook-main-fix-imports-autodocs-monorepo.js.mdx:19]: Are these the relevant properties? Should we highlight that?

File: docs/writing-docs/autodocs.md [hunk 11]
```diff
@@ -190,8 +234,50 @@ Out of the box, Storybook has a set of components that you can use to customize
 
 <!-- prettier-ignore-end -->
 
+<div class="aside">
+
+💡 This is not a Storybook issue but a breaking change introduced with MDX 2. For more information on this and other breaking changes, see our [MDX documentation](./mdx.md#breaking-changes).
+
+</div>
+
+## Troubleshooting
```
Comment (kylegach) [inline docs/writing-docs/autodocs.md:243]: So grateful we take the time to document these gotchas. 🙏

File: docs/writing-docs/mdx.md [hunk 12]
```diff
@@ -202,9 +203,9 @@ To help you transition to the new version, we've created a migration helper in o
 npx storybook@next automigrate mdx1to2
 ```
 
-## Documentation-only MDX
+## Extending documentation with MDX
```
Comment (kylegach) [inline docs/writing-docs/mdx.md:206]: I love the narrative of extending documentation.

But, we refer to this concept as "unattached" internally (and when describing the Meta doc block). Should we make that term official, so when searching it leads here?

File: docs/writing-docs/mdx.md [hunk 13]
```diff
@@ -254,36 +285,67 @@ import Changelog from '../CHANGELOG.md';
...
-This redirects to the **Canvas** tab of the story:
+<div class="aside">
+💡 Depending on the volume, you may be required to run the command multiple times to fix all the issues.
+</div>
+
+When it finishes, it will output the list of files causing issues. You can then use this information to fix the problems manually.
 
-```md
-[Go to specific story canvas](?path=/story/some--id)
+Additionally, if you're working with VSCode, you can add the [MDX extension](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx) and enable MDX experimental support by adding the following to your user settings:
```
Comment (kylegach) [inline docs/writing-docs/mdx.md:306]:
```suggestion
Additionally, if you're working with VSCode, you can add the [MDX extension](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx) and enable MDX experimental support for linting, type checking, and auto completion by adding the following to your user settings:
```

Imo, we should consider moving this somewhere more prominent, as it's a bit of a game changer.

**Comment by kylegach** [comment]
> Looked through the changes and it looks good! Is there an easy way to preview the changes on the site?

1. In `frontpage`, extract the docs for this branch: `yarn extract-monorepo-docs chore_auto_docs__docs_fixes`
2. Start the site: `GATSBY_SKIP_ADDON_PAGES=true yarn start`

/cc @JReinhold

File: docs/writing-docs/autodocs.md (hunk 7 above)
Comment (jonniebigodes) [reply docs/writing-docs/autodocs.md:92]: I left the `docs` parameter as a placeholder. Once the API is out, this will be updated to point at it. And to this point
> I'm pretty sure I've seen @tmeasday mention that you must provide a React component now and can no longer use MDX for this.
This is an escape hatch. You can write your custom templates and place them outside a directory that is not picked by the stories glob pattern. I left this intentionally to avoid having the community reach out to us in our communication channels and ask why they're forced to write React when they work with  Vue, Angular, etc. The directory choice (i.e., `src`) was intentional, as most framework generators already set it up. I'm aware of the fact that Web Components and a monorepo infrastructure work differently.

**Review by kylegach — DISMISSED** [review]
Nice work, overall! 👏 

I'm marking this as 'Request changes' due to the [`coreFrameworks` change](https://github.com/storybookjs/storybook/pull/21122#discussion_r1108876632), which must be resolved prior to merging.

File: docs/frameworks.js (hunk 3 above)
Comment (jonniebigodes) [reply docs/frameworks.js:2]: The same exact reason why Qwik was introduced. We're providing support for it. Shouldn't they be featured? if so, more than glad to update it.

File: docs/snippets/common/storybook-main-fix-imports-autodocs-monorepo.js.mdx (hunk 10 above)
Comment (jonniebigodes) [reply docs/snippets/common/storybook-main-fix-imports-autodocs-monorepo.js.mdx:19]: Yes, they are. I forgot about them. I'll update this to put a bit more emphasis on the required properties

File: docs/frameworks.js (hunk 3 above)
Comment (kylegach) [reply docs/frameworks.js:2]: I don't understand your comment about Qwik. That's not in the list of current or proposed (by this PR) core frameworks.

I suppose, technically, adding `nextjs` as an option in the FrameworkSelector menu would just result in all snippets falling back to `react`. Which in this specific case, would be fine.

But imagine adding `sveltekit` (which is just as valid as `nextjs`, imo)... all of those snippets would also fall back to `react`, which would be incorrect.

Adding more core frameworks requires `frontpage` updates, and probably more careful consideration.

File: docs/writing-docs/autodocs.md (hunk 7 above)
Comment (kylegach) [reply docs/writing-docs/autodocs.md:92]: > This is an escape hatch. You can write your custom templates and place them outside a directory that is not picked by the stories glob pattern.

I'm confused. Are you documenting how to override the docs template or are you illustrating an alternative approach where you don't use AutoDocs at all and instead provide a custom MDX file?

File: docs/frameworks.js (hunk 3 above)
Comment (jonniebigodes) [reply docs/frameworks.js:2]: Yes, NextJS defaults React. I've tested it locally. And I'm aware of Sveltekit and related frameworks. To keep this moving I'll remove it for now and follow up with you so that we can provide a better way to document it. I'll change it in a bit.

File: docs/writing-docs/autodocs.md (hunk 7 above)
Comment (tmeasday) [reply docs/writing-docs/autodocs.md:92]: > I'm pretty sure I've seen @tmeasday mention that you must provide a React component now and can no longer use MDX for this.

I'm sure you are correct about this, but I (nor @shilman) remember the exact reason for that, and the proof is in the pudding-it indeed does work!

> This is an escape hatch. You can write your custom templates and place them outside a directory that is not picked by the stories glob pattern.

There's one more detail -- you can use the `isTemplate` prop to `Meta` (which we should document too, I didn't because I thought it didn't work 🤦 ) to stop a `.mdx` file from being indexed:

```mdx
import { Meta, Title } from '@storybook/blocks';

## This is an autodocs template

<Meta isTemplate />

<Title />
```

File: docs/writing-docs/autodocs.md (hunk 8 above)
Comment (jonniebigodes) [reply docs/writing-docs/autodocs.md:128]: Can you elaborate on this a bit more? Are you suggesting moving the documentation to the MDX page starting at "Setup custom documentation" all the way to "Advanced configuration"? Is that what you're thinking?

File: docs/writing-docs/autodocs.md (hunk 9 above)
Comment (jonniebigodes) [reply docs/writing-docs/autodocs.md:158]: Addressed with the latest commit. Marking this as solved.

File: docs/writing-docs/autodocs.md (hunk 7 above)
Comment (jonniebigodes) [reply docs/writing-docs/autodocs.md:93]: I added some clarification on it with the latest commit.

File: docs/writing-docs/mdx.md [hunk 14]
```diff
@@ -254,36 +285,67 @@ import Changelog from '../CHANGELOG.md';
...
 -->
 
-## Linking to other stories and pages
+## Troubleshooting
 
-When writing MDX, you may want to provide links to other stories or documentation pages and sections. You can use the `path` query string.
+### The migration seems flaky and keeps failing
 
-Considering a story with ID `some--id`, this redirects to the **Docs** tab of the story:
+By default, running the [migration command](#automigration) will try and migrate all existing MDX files in your project according to the MDX 2 specification. However, this might not always be possible, and you might run into issues during the migration. To help you troubleshoot those issues, we've prepared some recommendations that might help you.
```
Comment (tmeasday) [inline docs/writing-docs/mdx.md:292]: I don't believe our migration tries to convert MDX1->2, or do I have that wrong @JReinhold @yannbf ?

Comment (tmeasday) [reply docs/writing-docs/mdx.md:292]: PS should this content be in our migration guide rather than the docs themselves? It doesn't see relevant to new users.

File: docs/writing-docs/mdx.md [hunk 15]
```diff
@@ -254,36 +285,67 @@ import Changelog from '../CHANGELOG.md';
...
 
-```md
-[Go to the conclusion of the documentation page](?path=/docs/some--id#conclusion)
-```
+### The MDX documentation doesn't render in my environment
 
-<div class="aside">
-💡 By applying this pattern with the Controls addon, all anchors will be ignored in Canvas based on how Storybook handles URLs to track the args values.
-</div>
+As Storybook relies heavily on MDX 2 to render documentation, there might be some cases where you cannot render it depending on your project setup. If that's the case, we support MDX 1 as a fallback. To enable it, you'll need to take some additional steps.
```
Comment (tmeasday) [inline docs/writing-docs/mdx.md:318]: I would pitch this more as: "if you don't want to migrate all your files to MDX2 right now..."

File: docs/writing-docs/autodocs.md (hunk 8 above)
Comment (kylegach) [reply docs/writing-docs/autodocs.md:141]: Yes, that's exactly what I'm suggesting. Everything in that section of content pertains to writing MDX documentation in general, not the automatic docs that Storybook provides.

I do think it'd be a good idea to point from here to there, though, explaining that if customizing the autoDocs page isn't enough, they can author their own custom MDX content.

File: docs/writing-docs/autodocs.md (hunk 7 above)
Comment (kylegach) [reply docs/writing-docs/autodocs.md:93]: I'll add `isTemplate` to the API reference. 👍

File: docs/writing-docs/mdx.md [hunk 16]
```diff
@@ -218,6 +219,36 @@ paths={[
 
 ![MDX docs only story](./mdx-documentation-only.png)
 
+### Linking to other stories and pages
+
+Another way to improve documentation is by linking to other stories and pages. Suppose you already have a component story with the following unique identifier, `some--id`, and you want to link it to your documentation page. In that case, you can use the `path` query string to redirect to the documentation entry related to the story:
+
+```md
+[Go to specific documentation page](?path=/docs/some--id)
```
[the remark this answers is not in the document]
Comment (JReinhold) [reply docs/writing-docs/mdx.md:227]: Hmm I don't know, but I can imagine linking between docs pages is a common pattern. What you're suggesting sounds like a new `<Link to={ButtonStories} />` block, which seems valuable to me.
Something to explore in upcoming releases for sure.

[the remark this answers is not in the document]
Comment (tmeasday) [reply docs/writing-docs/mdx.md:227]: OK, I guess we leave this for now, but I wonder if we should adding documentation that we probably plan to deprecate soon 🤷

[the remark this answers is not in the document]
Comment (jonniebigodes) [reply docs/writing-docs/mdx.md:227]: This was left in on a temporary basis until there's an API to address this to avoid the issues related to linking that continue to surface in the repo addressed. Once we have a stable API and a path moving forward, this section will be rewritten.

File: docs/writing-docs/mdx.md (hunk 14 above)
Comment (jonniebigodes) [reply docs/writing-docs/mdx.md:292]: I disagree. We should instead leave it here instead of the public notion doc acting as a migration guide, as this will affect not only new users but also existing users. As most folks will not see it, by adding it, we have the added bonus of discoverability via searching the documentation. Also, to which the CLI should also point here.
